compression_verifier.go

```diff
@@ -0,0 +1,224 @@
+		" not supported on table: " + e.table +
+		" for column: " + e.column
+}
+
+// CompressionVerifier provides the same functionality as the iterative verifier but also
+// provides support for manually verifying the payload of compressed columns that may
+// have different hashes for the same data
+type CompressionVerifier struct {
+	logger *logrus.Entry
+
+	// supportedAlgorithms provide O(1) lookup to check if a configured algorithm is supported
+	supportedAlgorithms map[string]struct{}
+
+	// tableCmpressions is represented as table[column][compression_algorithm]
+	tableCompressions map[string]map[string]string
+}
+
+// NewCompressionVerifier first checks the map for supported compression algorithms before
+// initializing and returning the initialized instance.
+func NewCompressionVerifier(tableCompressions map[string]map[string]string) (*CompressionVerifier, error) {
+	supportedAlgorithms := make(map[string]struct{})
+	supportedAlgorithms[CompressionSnappy] = struct{}{}
+
+	compressionVerifier := &CompressionVerifier{
+		logger:              logrus.WithField("tag", "compression_verifier"),
+		supportedAlgorithms: supportedAlgorithms,
+		tableCompressions:   tableCompressions,
+	}
+
+	if err := compressionVerifier.verifyConfiguredCompression(tableCompressions); err != nil {
+		return nil, err
+	}
+
+	return compressionVerifier, nil
+}
+
+func (c *CompressionVerifier) verifyConfiguredCompression(tableCompressions map[string]map[string]string) error {
+	for table, columns := range tableCompressions {
+		for column, algorithm := range columns {
+			// Ensure algorithm is uppercased
+			algorithm = strings.ToUpper(algorithm)
+			tableCompressions[table][column] = algorithm
+
+			if _, ok := c.supportedAlgorithms[algorithm]; !ok {
+				return &UnsupportedCompressionError{
+					table:     table,
+					column:    column,
+					algorithm: algorithm,
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// GetHashes compares the source data with the target data to ensure the integrity of the
+// data being copied.
+//
+// The CompressionVerifier method checks if the existing table contains compressed data
+// and will apply the decompression algorithm to the applicable columns if necessary.
+// After the columns are decompressed, the hashes of the data are used and verified
+// for comparison equality
+func (c *CompressionVerifier) GetHashes(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64, tableCompression map[string]string) (map[uint64][]byte, error) {
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := c.getRows(db, schema, table, pkColumn, columns, pks)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decompress applicable columns and hash the resulting column values for comparison
+	return c.decompressAndHash(table, columns, rows, tableCompression)
+}
+
+func (c *CompressionVerifier) getRows(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64) (*sql.Rows, error) {
```

Comment:
This function seems very generic and totally unrelated to compression. Feels like maybe this should be on some other struct and not on `CompressionVerifier`? Do we have this code elsewhere already from which we could reuse it?